### PR TITLE
Add CSM 2D Drawing API

### PR DESCRIPTION
**Pull request by v-rob**
This is the cleaned up and bugfixed drawing portion of #10524.  Essentially, it adds a `Texture` class that allows drawing to custom textures and the screen, specifically pixels, filled and gradiented rectangles, textures, text, and items.  The `Texture` class has no support for in-game textures (that's for another PR).  In addition, it introduces a new versatile rectangle and positioning system which likely will be used in the actual formspec replacement API as well.

`client_lua_api.txt` is the best place to find the actual documentation

## To do

This PR is a Work in Progress.  There are a few small bugs, and fonts still need reworking.

Notably, drawing models is absent from this PR.  This is because I plan to add a good model system with support for multiple models in a scene (allowing e.g. the player model with armor on and a backpack), but that's too much for this PR, which is big enough already.  A way to draw lines, filled polygons, and other shapes will come in the 3D drawing PR as well.

## Discussion points

* Currently, positions are based on 0, not 1 like Lua, i.e. `{x=0, y=0}` is the top left, not `{x=1, y=1}`.  Should it be based on 1 to match Lua?

## How to test

This PR creates a new bundled CSM, `devtest_csm`, since there is no way to actually unit test this in the server-side `devtest`.  To test, enable this CSM and use the command `test_draw`.  The param `off` turns off the drawing, `window` turns on drawing to the Minetest window, `texture` draws it to a custom texture which is then drawn to the screen, `toggle_fill` turns filling the screen/texture with a color before drawing on and off, `dump_driver` prints the current video driver specs, and `test_ref` tests texture sizes via texture references.

For instance, here is what drawing to the window with filling on should look like:
screenshot_20210112_175347

### CODE REVIEW

File: doc/client_lua_api.txt [hunk 1]
```diff
@@ -1152,6 +1277,114 @@ Can be obtained via `minetest.get_meta(pos)`.
...
+* `font`: What font to use. Can be "normal" or "mono". Default "normal". If the fallback font
+  is used for the current locale, this option will be overridden.
+* `size`: What size the font should be. The default is user-set.
+	* **Note**: It is recommended to make the size relative to the user-set font size which
+	  can be gotten with `minetest.get_font_size`.
+* `bold`: Whether the text should be bold. Default false.
+* `italic`: Whether the text should be italicized. Default false.
+* `underline`: Whether the text should be underlined. Default false.
+* `strikethrough`: Whether the text should be crossed out. Default false.
+
```
Comment (kilbith) [inline doc/client_lua_api.txt:1425]: `highlight` please?

Comment (v-rob) [reply doc/client_lua_api.txt:1387]: I considered it, but concluded that it's not especially necessary.  It's easy enough to add highlight by using a `minetest.get_text_size` and then `draw_rect` the highlight before drawing the text.  On the other hand, italics, underline, and suchlike can't be done from Lua (underline requires knowledge of the font ascenders), hence their inclusion here.

So, `highlight` can be trivially added if wanted (and it certainly will be in the UI even if not in the drawing API), but it doesn't add any functionality that can't already be done easily enough.  The same goes for why there isn't align/valign functionality built in.

Comment (kilbith) [reply doc/client_lua_api.txt:1425]: > It's easy enough to add highlight by using a minetest.get_text_size and then draw_rect the highlight before drawing the text

You have to manually find the position (not in terms of character index but coordinates) of the substring, that's painful. Not sure everyone can perfectly place their rects under the substrings. Technically, you can also draw a rect for `underline` and `strikethrough`.

**Comment by kilbith** [comment]
By the way, `draw_text` is _far_ for being able to do what `hypertext` can do with its tag system. If one wants to switch to a different font style in the middle of a text, they have to call `draw_text` _another time_ with a different style param, _properly position_ the string according to the previously written text, and then call `draw_text` _again_ back to a normal style, and properly position it _again_ (with little guarantee it will be perfect).

Do you realize how painful it is?

**Comment by v-rob** [comment]
> By the way, `draw_text` is _far_ for being able to do what `hypertext` can do with its tag system.

Very true.  In fact, it can't even do what `label[]` can do -- `draw_text` strips color escape codes as well.  This is actually intentional for two reasons.  The reason is that this API is meant to be lower level, used underneath another system (specifically the GUI), so unnecessary features are kept out.  If the GUI can implement it, then the drawing API doesn't.

The second reason is that not even `hypertext[]` can do everything.  If someone wants to have a 1 pixel border around their text with an orange background with two pixels of padding, then that person's out of luck if they're using `hypertext[]`.  (This isn't meant to insult `hypertext[]`, just showing my point that nothing can do everything.)  A lot of these features are subjective, and trying to add them all to one function can't cut it.  That's why these other types of features are left to the system using this API.  Even if I allowed some extra features (like the color escape sequences), that would just make the API harder to use for advanced behaviour (such as if the user wants to implement their own way to change color mid-string without having to work around the existing escape sequences).

Rest assured, the GUI will without a doubt have an element to match `hypertext[]`'s abilities; it's just that that ability is the GUI's responsibility, not the drawing API's.

File: doc/client_lua_api.txt (hunk 1 above)
Comment (v-rob) [reply doc/client_lua_api.txt:1387]: `underline` and `strikethrough` require the usage of `CGUITTFont::getAscender`, so they actually can't be done as well Lua-side, not without exposing `getAscender`.  That's why they're included.

**Comment by kilbith** [comment]
Can the `underline` and `strikethrough` code be backported to `GUIHyperText` btw?

**Comment by v-rob** [comment]
> Can the `underline` and `strikethrough` code be backported to `GUIHyperText` btw?

Truth be told, that's where I got the `underline` code; it's used for the links, but I don't think it's available to the user elsewhere.  `strikethrough` would be trivially implementable.

**Comment by Andrey2470T** [comment]
If this PR allows to draw on textures, could it particularly implement such feature as the mirror reflections on nodes surfaces?

**Comment by v-rob** [comment]
> If this PR allows to draw on textures, could it particularly implement such feature as the mirror reflections on nodes surfaces?

This PR doesn't allow the textures to be used in game currently, but only with the drawing API.  Regardless, mirror reflections are much better handled by shaders.

**Comment by v-rob** [comment]
Three things:

1. I've decided that the best way to draw lines, filled polygons, and other shapes is probably to use a `SMeshBuffer` because software drawing is a bad idea and I have no idea how to do shaders (which probably would only work on OpenGL anyway), so hardware accelerated 3D sounds like the best option to me.  Therefore, it will go in the later 3D PR.
2. If anyone knows how to make simple shaders that can do effects like blur, apply noise, or pixelate things, those could be really nice to make really good-looking GUIs (and special in-game effects for that matter).
3. The font API needs some work, and I will be fixing it up soon:
    * Underline and strikethrough are too hardcoded for my taste, so I think I'll add three new keys to the return value of `minetest.get_text_size`: `underline_y`, the Y position of the top of the underline, `strike_y`, the Y position of the top of the strikethrough, and `line_height`, the height of the underline/strikethrough.  This will allow mods to have custom line colors independent of the font color as well as other designs like dotted and dashed lines without the engine having to provide every possibility itself.  Therefore, the `underline` and `strike` options would be removed from FontSpec as well.
    * Font shadows are also hardcoded.  Since it's super easy for a mod to draw a font shadow itself, I think that the font shadow will either a) be disabled in `draw_font` automatically or b) it can be optionally disabled with FontSpec.  Hardcoded font shadows are not versatile even though their offsets and color can be modified because underlines and suchlike can't be drawn for them properly.
    * I know that the font needs some user input, but things like underlines, font shadows, and the like are better decided by a GUI  designer to work with the color scheme and suchlike properly.  Allowing the user to make the first decisions will more likely mess GUIs up than help them.  Font family and font size, of course, must be left for the user to decide since that's an accessibility concern.  I also know that making the font API barebones makes it more difficult to use, but it's going to be inside the GUI API, which will have many more options.

**Comment by rubenwardy** [comment]
> If anyone knows how to make simple shaders that can do effects like blur, apply noise, or pixelate things, those could be really nice to make really good-looking GUIs (and special in-game effects for that matter).

I've written blur shaders, it's not too bad - you just need to apply it twice, which results post processing

**Comment by v-rob** [comment]
OK, a lot of changes.  I'll summarize some of the bigger ones:
* Fonts now have their own class, `Font`, which is a much better and more centralized place for them.  True Freetype metrics are used when possible, so underlines and stuff are 100% correct now.  Underlines and strikethroughs must now be drawn manually, but that means they can support e.g. dotted or dashed lines.
* 9-slice images have been removed from `draw_texture` since they can be implemented in the GUI library just as well (and better).
* `draw_item` has been removed.  It will be reimplemented in the 3D PR in a much more versatile manner, such as with an `ItemSceneNode` or similar.
* Hacks around Irrlicht's `setViewPort` bug have been fixed at the source
* Quite a few miscellaneous bugs have been fixed.

Still, it's not done, I'm sorry to say.  There are a few more known bugs, and I want a few more features, namely software pixel reading and writing, reading and writing textures to and from strings, and potentially making `minetest.window` readable via a render texture.

**Comment by v-rob** [comment]
Closing because the GUI's being done from another angle.

It's not being abandoned!  I will still make Lua drawing APIs, but I need to do them a different way now.  The way I've been doing them isn't extensible enough to work for more than the GUI, and since the GUI doesn't _need_ them anymore, that's not a good way to be doing things.  They need to be done in a more extensible way so they can be used anywhere in the engine, which means that I have to make some of my changes in Irrlicht's `IVideoDriver` and `ITexture` directly instead of here in Minetest.

**Comment by alwayshopeless** [comment]
Closing this PR was a mistake.
In fact, the existence of the Lua-centered Canvas API is entirely justified, regardless of how the new GUI is implemented.

This API opens the way to creating minigames, maps, visualizing achievement trees, etc.
Please reopen it if you don't plan to include it as part of the new GUI API.

**Pull request closed without merging**